src/main/java/htsjdk/samtools/HtsgetBAMFileReader.java

```diff
@@ -0,0 +1,561 @@
+                        : currInterval.overlaps(rec) && (prevInterval == null || !prevInterval.overlaps(rec));
+                }
+
+                private boolean acceptUnmappedRecord(final SAMRecord rec) {
+                    final int start = rec.getStart();
+                    final boolean matchesCurrInterval =
+                        rec.contigsMatch(currInterval) &&
+                            CoordMath.encloses(currInterval.getStart(), currInterval.getEnd(), start, start);
+                    final boolean matchesPrevInterval =
+                        prevInterval != null &&
+                            rec.contigsMatch(prevInterval) &&
+                            CoordMath.encloses(prevInterval.getStart(), prevInterval.getEnd(), start, start);
+                    return matchesCurrInterval && !matchesPrevInterval;
+                }
+            };
+            this.currentIterator = new FilteringSamIterator(this.iterators.next().get(), filter, false);
+            this.currentIntervalIndex++;
+        }
+    }
+
+    public static class BAMQueryFilteringIterator implements CloseableIterator<SAMRecord> {
+        /**
+         * The wrapped iterator.
+         */
+        private final CloseableIterator<SAMRecord> wrappedIterator;
+        /**
+         * The next record to be returned.  Will be null if no such record exists.
+         */
+        private SAMRecord mNextRecord;
+        private final BAMIteratorFilter iteratorFilter;
+
+        public BAMQueryFilteringIterator(final CloseableIterator<SAMRecord> iterator, final BAMIteratorFilter iteratorFilter) {
+            this.wrappedIterator = iterator;
+            this.iteratorFilter = iteratorFilter;
+            this.mNextRecord = this.advance();
+        }
+
+        /**
+         * Returns true if a next element exists; false otherwise.
+         */
+        @Override
+        public boolean hasNext() {
+            return this.mNextRecord != null;
+        }
+
+        /**
+         * Gets the next record from the given iterator.
+         *
+         * @return The next SAM record in the iterator.
+         */
+        @Override
+        public SAMRecord next() {
+            if (!this.hasNext()) {
+                throw new NoSuchElementException("BAMQueryFilteringIterator: no next element available");
+            }
+            final SAMRecord currentRead = this.mNextRecord;
+            this.mNextRecord = this.advance();
+            return currentRead;
+        }
+
+        SAMRecord advance() {
+            while (true) {
+                // Pull next record from stream
+                if (!this.wrappedIterator.hasNext()) {
+                    return null;
+                }
+
+                final SAMRecord record = this.wrappedIterator.next();
+                switch (this.iteratorFilter.compareToFilter(record)) {
+                    case MATCHES_FILTER:
```

Comment:
in case statements like this I actually think putting the case and code on the same line is fine.